src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+    }
+    this->i++;
+    return true;
+}
+
+static void
+filterRecord(WalFilterState *const this)
+{
+    // In the case of overwrite contrecord, we do not need to try to filter it, since the record may not have a body at all.
+    if (this->got_len != this->record->xl_tot_len)
+    {
+        return;
+    }
+
+    RelFileNode *node;
+    // Pass through the records that are related to the system catalog or system databases (template1, template0 and postgres)
+    if (getRelFileNodeGPDB6(this->record, &node))
+    {
+        bool isPassTheFilter = false;
+
+        // isRelationNeeded can allocate memory on the first call. Therefore, we switch the context.
+        MEM_CONTEXT_OBJ_BEGIN(this)
+        isPassTheFilter = isRelationNeeded(node->dbNode, node->spcNode, node->relNode);
+        MEM_CONTEXT_OBJ_END();
+
+        if (isPassTheFilter)
+        {
+            return;
+        }
+
+        this->record->xl_rmid = RESOURCE_MANAGER_XLOG_ID;
+        this->record->xl_info = NOOP_XLOG;
+        uint32_t crc = crc32cInit();
+        crc = crc32cComp(crc, (unsigned char *) XLogRecGetData(this->record), this->record->xl_len);
+        crc = crc32cComp(crc, (unsigned char *) this->record, offsetof(XLogRecord, xl_crc));
+        crc = crc32cFinish(crc);
+
+        this->record->xl_crc = crc;
+    }
+}
+
+static void
+writeRecord(WalFilterState *const this, Buffer *const output)
+{
+    uint32_t header_i = 0;
+    if (this->total_offset % this->walPageSize == 0)
+    {
+        ASSERT(!lstEmpty(this->headers));
+        const XLogPageHeaderData *const header = lstGet(this->headers, header_i);
+        const size_t to_write = XLogPageHeaderSize(header);
+        checkOutputSize(output, to_write);
+        memcpy(bufRemainsPtr(output), header, to_write);
+        bufUsedInc(output, to_write);
+        this->total_offset += to_write;
+        header_i++;
+    }
+
+    size_t wrote = 0;
+    while (this->got_len != wrote)
+    {
+        const size_t space_on_page = this->walPageSize - bufUsed(output) % this->walPageSize;
+        size_t to_write = Min(space_on_page, this->got_len - wrote);
+        checkOutputSize(output, to_write);
+
+        memcpy(bufRemainsPtr(output), ((char *) this->record) + wrote, to_write);
+        wrote += to_write;
+
+        bufUsedInc(output, to_write);
+
+        // write header
+        if (header_i < lstSize(this->headers))
+        {
+            to_write = SizeOfXLogShortPHD;
+            checkOutputSize(output, to_write);
+            memcpy(bufRemainsPtr(output), lstGet(this->headers, header_i), to_write);
+            bufUsedInc(output, to_write);
+            this->total_offset += to_write;
+            header_i++;
+        }
+    }
+
+    const size_t align_size = MAXALIGN(this->got_len) - this->got_len;
+    checkOutputSize(output, align_size);
+    memset(bufRemainsPtr(output), 0, align_size);
+    bufUsedInc(output, align_size);
+
+    this->total_offset += this->tot_len;
+    this->got_len = 0;
+}
```

Comment:
Is it possible to simplify this function to write only what is read from the buffer, and not the parsed record? Except, of course, the nopped record.